textdata_preprocessor.py

```diff
@@ -0,0 +1,253 @@
+            else:
+                end = len(input_pack.text)
+
+        if end > begin:
+            findings = Findings(input_pack, begin, end)
+            findings.has_content = True
+        else:
+            findings = Findings(input_pack, end, end)
+            findings.has_content = False
+
+        counter = self.resources.get('counter')
+        if findings.has_content and counter is not None:
+            # Update the vocabulary counter
+            text = input_pack.text[begin: end].replace(',', '').replace('.', '')
+            counter.update(text.split(' '))
+
+
+class ImpressionExtractor(PackProcessor):
+    r"""A processor to extract the Impression session in the medical report.
+    """
+
+    def _process(self, input_pack: DataPack):
+        findings_ind = input_pack.text.find("FINDINGS")
+        impression_ind = input_pack.text.find("IMPRESSION")
+        if impression_ind == -1:
+            begin = 0
+            end = 0
+        else:
+            begin = impression_ind + len("IMPRESSION") + 1
+            if findings_ind != -1 and impression_ind < findings_ind:
+                end = findings_ind - 1
+            else:
+                end = len(input_pack.text)
+
+        if end > begin:
+            impression = Impression(input_pack, begin, end)
+            impression.has_content = True
+        else:
+            impression = Impression(input_pack, end, end)
+            impression.has_content = False
+
+        counter = self.resources.get('counter')
+        if impression.has_content and counter is not None:
+            # Update the vocabulary counter
+            text = input_pack.text[begin: end].replace(',', '').replace('.', '')
+            counter.update(text.split(' '))
+
+
+class NonAlphaTokenRemover(MultiPackProcessor):
+    r"""A class of non alpha token remover that requires a nltk tokenizer in the
+    upstream of the pipeline. The modified text would be added to a new pack.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.in_pack_name = 'default'
+        self.out_pack_name = 'result'
+
+    def _process(self, input_pack: MultiPack):
+        pack_name_string = input_pack.get_pack(self.in_pack_name).pack_name
+
+        token_entries = list(input_pack.get_pack(
+            self.in_pack_name).get(entry_type=Token))
+        token_texts = [token.text for token in token_entries]
+        words = [word for word in token_texts
+                 if word.isalpha() or word[:-1].isalpha()
+                 or word == '.' or osp.isfile(word)]
+
+        if words[-1] == '.':
+            # Combine the last '.' with the final word
+            words = words[:-1]
+            words[-1] += '.'
+
+        pack = input_pack.add_pack(self.out_pack_name)
+        result = ' '.join(words)
+
+        pack.set_text(text=result)
+        filepath = FilePath(pack)
+
+        if pack_name_string is not None:
+            pack_name_string = pack_name_string.replace('.jpg', '')
```

Comment:
Looks like the system will not work without the img path. So we may raise exception here if we cannot find the pack name string (which is the pack_name of the input_pack)

Reply:
Added codes to raise the exception when the corresponding parent image file name cannot be found.